src/data/bucket/line_attributes_ext.js

```diff
@@ -0,0 +1,11 @@
+// @flow
+import {createLayout} from '../../util/struct_array';
+
+const lineLayoutAttributesExt = createLayout([
+    {name: 'a_line_progress', components: 1, type: 'Float32'},
+    {name: 'a_line_clip', components: 1, type: 'Float32'},
+    {name: 'a_split_index', components: 1, type: 'Float32'},
+]);
```

Comment:
This makes vertices three times bigger with 24 bytes per vertex instead of 8 bytes per vertex when the clip stats are present. Right now this would only affect geojson layers, right?

Do `a_line_progress` and `a_line_clip` need to be stored separately or can we store `a_line_progress / a_line_clip` as a single attribute?

Reply:
This extension is an optional 12 bytes per vertex which only affects geojson layers that request for line metrics, memory is the main reason to have them in a separate buffer. It's a total of 8 (LineLayoutArray) + 12 (LineExtLayoutArray) for the geojson + line metrics code path.

If we want to maximize accuracy I don't think we would be able to lower precision here, it would be quite destructive for what we're trying to improve. But there are other ways we can save on memory by using vertex divisor. Clips and split_index replicate across the geometry so it would be a good candidate for instancing optimization: https://developer.mozilla.org/en-US/docs/Web/API/ANGLE_instanced_arrays.

Reply:
As for using a single attribute I have no preference, we could directly bake it in a single if that makes more sense?
```
{name: 'a_line_metrics', components: 2, type: 'Float32'},
{name: 'a_split_index', components: 1, type: 'Float32'},
```

Reply:
By baking into a single attribute I meant doing the division of `a_line_progress / a_line_clip` before adding it to the buffer rather than in the shader. Would this work?

Reply:
Oh right sorry I misread. I think that should work, we're in the range 0..1 and should stay within that range even by doing the division beforehand, so it should be safe to do so. I'll update for that.

Reply:
Added in https://github.com/mapbox/mapbox-gl-js/pull/9694/commits/2adb186f726e05fead7fb855b688de22ecaa49fe as well as some other improvements.